File: cmd/tags.go
// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2024-present Datadog, Inc.

package cmd

import (
	"fmt"

	"github.com/DataDog/datadog-api-client-go/v2/api/datadogV1"
	"github.com/spf13/cobra"
)

var tagsCmd = &cobra.Command{
	Use:   "tags",
	Short: "Manage host tags",
	Long: `Manage tags for hosts in your infrastructure.

Tags provide metadata about your hosts and help organize and filter
your infrastructure.

CAPABILITIES:
  • List all host tags
  • Get tags for a specific host
  • Add tags to a host
  • Update host tags
  • Remove tags from a host

EXAMPLES:
  # List all host tags
  pup tags list

  # Get tags for a host
  pup tags get my-host

  # Add tags to a host
  pup tags add my-host env:prod team:backend

AUTHENTICATION:
  Requires either OAuth2 authentication or API keys.`,
}

var tagsListCmd = &cobra.Command{
	Use:   "list",
	Short: "List all host tags",
	RunE:  runTagsList,
}

var tagsGetCmd = &cobra.Command{
	Use:   "get [hostname]",
	Short: "Get tags for a host",
	Args:  cobra.ExactArgs(1),
	RunE:  runTagsGet,
}

var tagsAddCmd = &cobra.Command{
	Use:   "add [hostname] [tags...]",
	Short: "Add tags to a host",
	Args:  cobra.MinimumNArgs(2),
	RunE:  runTagsAdd,
}

var tagsUpdateCmd = &cobra.Command{
	Use:   "update [hostname] [tags...]",
	Short: "Update host tags",
	Args:  cobra.MinimumNArgs(2),
	RunE:  runTagsUpdate,
}

var tagsDeleteCmd = &cobra.Command{
	Use:   "delete [hostname]",
	Short: "Delete all tags from a host",
	Args:  cobra.ExactArgs(1),
	RunE:  runTagsDelete,
}

func init() {
	tagsCmd.AddCommand(tagsListCmd, tagsGetCmd, tagsAddCmd, tagsUpdateCmd, tagsDeleteCmd)
}

func runTagsList(cmd *cobra.Command, args []string) error {
	client, err := getClient()
	if err != nil {
		return err
	}

	api := datadogV1.NewTagsApi(client.V1())
	resp, r, err := api.ListHostTags(client.Context())
	if err != nil {
		if r != nil {
			return fmt.Errorf("failed to list host tags: %w (status: %d)", err, r.StatusCode)
		}
		return fmt.Errorf("failed to list host tags: %w", err)
	}

	return formatAndPrint(resp, nil)
}

func runTagsGet(cmd *cobra.Command, args []string) error {
	client, err := getClient()
	if err != nil {
		return err
	}

	hostname := args[0]
	api := datadogV1.NewTagsApi(client.V1())
	resp, r, err := api.GetHostTags(client.Context(), hostname)
	if err != nil {
		if r != nil {
			return fmt.Errorf("failed to get host tags: %w (status: %d)", err, r.StatusCode)
		}
		return fmt.Errorf("failed to get host tags: %w", err)
	}

	return formatAndPrint(resp, nil)
}

func runTagsAdd(cmd *cobra.Command, args []string) error {
	client, err := getClient()
	if err != nil {
		return err
	}

	hostname := args[0]
	tags := args[1:]

	api := datadogV1.NewTagsApi(client.V1())
	body := datadogV1.HostTags{
		Tags: tags,
	}

	resp, r, err := api.CreateHostTags(client.Context(), hostname, body)
	if err != nil {
		if r != nil {
			return fmt.Errorf("failed to add host tags: %w (status: %d)", err, r.StatusCode)
		}
		return fmt.Errorf("failed to add host tags: %w", err)
	}

	return formatAndPrint(resp, nil)
}

func runTagsUpdate(cmd *cobra.Command, args []string) error {
	client, err := getClient()
	if err != nil {
		return err
	}

	hostname := args[0]
	tags := args[1:]

	api := datadogV1.NewTagsApi(client.V1())
	body := datadogV1.HostTags{
		Tags: tags,
	}

	resp, r, err := api.UpdateHostTags(client.Context(), hostname, body)
	if err != nil {
		if r != nil {
			return fmt.Errorf("failed to update host tags: %w (status: %d)", err, r.StatusCode)
		}
		return fmt.Errorf("failed to update host tags: %w", err)
	}

	return formatAndPrint(resp, nil)
}

func runTagsDelete(cmd *cobra.Command, args []string) error {
	client, err := getClient()
	if err != nil {
		return err
	}

	hostname := args[0]
	if !cfg.AutoApprove {
		fmt.Printf("⚠️  WARNING: This will delete all tags from host %s\n", hostname)
		fmt.Print("Are you sure you want to continue? (y/N): ")
		var response string
		if _, err := fmt.Scanln(&response); err != nil {
			// User cancelled or error reading input
			fmt.Println("\nOperation cancelled")
			return nil
		}
		if response != "y" && response != "Y" {
			fmt.Println("Operation cancelled")
			return nil
		}
	}

	api := datadogV1.NewTagsApi(client.V1())
	r, err := api.DeleteHostTags(client.Context(), hostname)
	if err != nil {
		if r != nil {
			return fmt.Errorf("failed to delete host tags: %w (status: %d)", err, r.StatusCode)
		}
		return fmt.Errorf("failed to delete host tags: %w", err)
	}

	fmt.Printf("Successfully deleted all tags from host %s\n", hostname)
	return nil
}

File: src/commands/tags.rs
use anyhow::Result;
use datadog_api_client::datadogV1::api_tags::{
    CreateHostTagsOptionalParams, DeleteHostTagsOptionalParams, GetHostTagsOptionalParams,
    ListHostTagsOptionalParams, TagsAPI, UpdateHostTagsOptionalParams,
};
use datadog_api_client::datadogV1::model::HostTags;

use crate::client;
use crate::config::Config;
use crate::formatter;

pub async fn list(cfg: &Config) -> Result<()> {
    let dd_cfg = client::make_dd_config(cfg);
    let api = match client::make_bearer_client(cfg) {
        Some(c) => TagsAPI::with_client_and_config(dd_cfg, c),
        None => TagsAPI::with_config(dd_cfg),
    };
    let resp = api
        .list_host_tags(ListHostTagsOptionalParams::default())
        .await
        .map_err(|e| anyhow::anyhow!("failed to list tags: {e:?}"))?;
    formatter::output(cfg, &resp)
}

pub async fn get(cfg: &Config, hostname: &str) -> Result<()> {
    let dd_cfg = client::make_dd_config(cfg);
    let api = match client::make_bearer_client(cfg) {
        Some(c) => TagsAPI::with_client_and_config(dd_cfg, c),
        None => TagsAPI::with_config(dd_cfg),
    };
    let resp = api
        .get_host_tags(hostname.to_string(), GetHostTagsOptionalParams::default())
        .await
        .map_err(|e| anyhow::anyhow!("failed to get tags: {e:?}"))?;
    formatter::output(cfg, &resp)
}

pub async fn add(cfg: &Config, hostname: &str, tags: Vec<String>) -> Result<()> {
    let dd_cfg = client::make_dd_config(cfg);
    let api = match client::make_bearer_client(cfg) {
        Some(c) => TagsAPI::with_client_and_config(dd_cfg, c),
        None => TagsAPI::with_config(dd_cfg),
    };
    let body = HostTags::new().tags(tags);
    let resp = api
        .create_host_tags(
            hostname.to_string(),
            body,
            CreateHostTagsOptionalParams::default(),
        )
        .await
        .map_err(|e| anyhow::anyhow!("failed to add tags: {e:?}"))?;
    formatter::output(cfg, &resp)
}

pub async fn update(cfg: &Config, hostname: &str, tags: Vec<String>) -> Result<()> {
    let dd_cfg = client::make_dd_config(cfg);
    let api = match client::make_bearer_client(cfg) {
        Some(c) => TagsAPI::with_client_and_config(dd_cfg, c),
        None => TagsAPI::with_config(dd_cfg),
    };
    let body = HostTags::new().tags(tags);
    let resp = api
        .update_host_tags(
            hostname.to_string(),
            body,
            UpdateHostTagsOptionalParams::default(),
        )
        .await
        .map_err(|e| anyhow::anyhow!("failed to update tags: {e:?}"))?;
    formatter::output(cfg, &resp)
}

pub async fn delete(cfg: &Config, hostname: &str) -> Result<()> {
    let dd_cfg = client::make_dd_config(cfg);
    let api = match client::make_bearer_client(cfg) {
        Some(c) => TagsAPI::with_client_and_config(dd_cfg, c),
        None => TagsAPI::with_config(dd_cfg),
    };
    api.delete_host_tags(
        hostname.to_string(),
        DeleteHostTagsOptionalParams::default(),
    )
    .await
    .map_err(|e| anyhow::anyhow!("failed to delete tags: {e:?}"))?;
    println!("Successfully deleted all tags from host {hostname}");
    Ok(())
}
